i6_models/config.py

```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
 
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class SubassemblyFactory:
```

Comment:
```suggestion
class SubassemblyFactory(Generic[ConfigType, ModuleType]):
```

Reply:
We can add this, but I have to admit even after reading up on it I dont fully understand the benefits over just having the parameters typed. Also is the order within the generic relevant? Then the types would need to be swapped right?

Reply:
If you want to use generic types inside a class you usually declare it as a `Generic` depending on these generic types. Having this "connects" the generic types used inside. For example `ModuleType` used in `module_class` is the same as the `ModuleType` returned by the `__call__` function. So for example if you set `module_class = ConformerEncoderV1` somewhere type checkers will know that the `__call__` method will return a `ConformerEncoderV1` instance.

> Also is the order within the generic relevant?

No, I don't believe the order is relevant.

Reply:
After reading a bit myself, I think the advantage is that we could specify which sort of subassemblies we would like to allow. e.g.

```
# define a metaclass for Norms
@dataclass
class NormAssemblyConfiguration(ModelConfiguration):
  pass

class NormAssembly(torch.nn.Module):
  pass

# define some norms
@dataclass
class BatchNormAssemblyConfiguration(NormAssemblyConfiguration):
  pass

class BatchNormAssembly(NormAssembly):
  pass

@dataclass
class LayerNormAssemblyConfiguration(NormAssemblyConfiguration):
  pass

class LayerNormAssembly(NormAssembly):
  pass

# some non norm related Assembly
class UnrelatedAssemblyConfiguration(ModelConfiguration):
  pass

class UnrelatedAssembly(torch.nn.Module):
  pass

# define a custom AssemblyConfig, that should accept any norm, but only norms
@dataclass
class MyAssemblyConfig(ModelConfiguration):
   norm: SubassemblyFactory[NormAssemblyConfiguration, NormAssembly]

# test the new config class
c1 = MyAssemblyConfig(
    norm=SubassemblyFactory(
        module_class=BatchNormAssembly,
        cfg=BatchNormAssemblyConfiguration()
    )
)  # correct typing

c2 = MyAssemblyConfig(
    norm=SubassemblyFactory(
        module_class=UnrelatedAssembly,
        cfg=UnrelatedAssemblyConfiguration()
    )
)  # valid python, but should give typing error
```
see my test here
![image](https://github.com/rwth-i6/i6_models/assets/23455405/3a43f230-2bc9-467c-9436-3c491dce9aac)


> Also is the order within the generic relevant?

you should match the order when typing your own configs. In the example above
```
norm: SubassemblyFactory[NormAssemblyConfiguration, NormAssembly]  # same order
norm: SubassemblyFactory[NormAssembly, NormAssemblyConfiguration]  # wrong order
```

> Then the types would need to be swapped right?

this is a matter of taste. I would agree, but there is no technical reason the order here should match the order of arguments in the SubassemblyFactory

Reply:
> If you want to use generic types inside a class you usually declare it as a Generic depending on these generic types. Having this "connects" the generic types used inside. 

I thought the fact that `ModuleType` is a `TypeVar` already did this, but apparently not.
![image](https://github.com/rwth-i6/i6_models/assets/23455405/92d830d4-fcc9-447e-9776-c1829c13de77)
(Type variable "ModuleType" has no meaning in this context)

However, if you move away from dataclasses, then you do not need to make the class a Generic to use TypeVars to enforce consistency:
![image](https://github.com/rwth-i6/i6_models/assets/23455405/6f65d921-8e0a-48a6-bf55-b24526142a7f)
("LayerNormAssemblyConfiguration" is incompatible with "BatchNormAssemblyConfiguration")

Reply:
> However, if you move away from dataclasses, then you do not need to make the class a Generic to use TypeVars to enforce consistency

This example doesn't work however when you add `def __call__(self) -> ModuleType:` because the system doesn't know that the `ModuleType` from `__init__` is the same `ModuleType` as in `__call__` if you don't inherit from `Generic`.

Reply:
> This example doesn't work however when  [...]

Indeed it does not. 😅 
Thanks for the clarification and counterexample.